stored_responses/stored_responses.go

```diff
@@ -0,0 +1,184 @@
+	// 2) stored bid responses: imp id to bidder to stored response id
+	impBiddersWithBidResponseIDs := make(map[string]map[string]string)
+	// 3) imp id to stored resp id
+	impAuctionResponseIDs := make(map[string]string)
+
+	for index, impData := range impInfo {
+		impId, err := jsonparser.GetString(impData.Imp, "id")
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("request.imp[%d] missing required field: \"id\"", index)
+		}
+
+		if impData.ImpExtPrebid.StoredAuctionResponse != nil {
+			if len(impData.ImpExtPrebid.StoredAuctionResponse.ID) == 0 {
+				return nil, nil, nil, fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)
+			}
+			allStoredResponseIDs = append(allStoredResponseIDs, impData.ImpExtPrebid.StoredAuctionResponse.ID)
+
+			impAuctionResponseIDs[impId] = impData.ImpExtPrebid.StoredAuctionResponse.ID
+
+		}
+		if len(impData.ImpExtPrebid.StoredBidResponse) > 0 {
+
+			bidderStoredRespId := make(map[string]string)
+			for _, bidderResp := range impData.ImpExtPrebid.StoredBidResponse {
+				if len(bidderResp.ID) == 0 || len(bidderResp.Bidder) == 0 {
+					return nil, nil, nil, fmt.Errorf("request.imp[%d] has ext.prebid.storedbidresponse specified, but \"id\" or/and \"bidder\" fields are missing ", index)
+				}
+				//check if bidder is valid/exists
+				if _, isValid := bidderMap[bidderResp.Bidder]; !isValid {
+					return nil, nil, nil, fmt.Errorf("request.imp[impId: %s].ext contains unknown bidder: %s. Did you forget an alias in request.ext.prebid.aliases?", impId, bidderResp.Bidder)
+				}
+				// bidder is unique per one bid stored response
+				// if more than one bidder specified the last defined bidder id will take precedence
+				bidderStoredRespId[bidderResp.Bidder] = bidderResp.ID
+				impBiddersWithBidResponseIDs[impId] = bidderStoredRespId
+				//storedAuctionResponseIds are not unique, but fetch will return single data for repeated ids
+				allStoredResponseIDs = append(allStoredResponseIDs, bidderResp.ID)
+			}
+		}
+	}
+	return allStoredResponseIDs, impBiddersWithBidResponseIDs, impAuctionResponseIDs, nil
+}
+
+// ProcessStoredResponses takes the incoming request as JSON with any
+// stored requests/imps already merged into it, scans it to find any stored auction response ids
+// in the request/imps and produces a map of imp IDs to stored auction responses.
+// Note that processStoredResponses must be called after processStoredRequests
+// because stored imps and stored requests can contain stored auction responses
+// so the stored requests/imps have to be merged into the incoming request prior to processing stored auction responses.
+func ProcessStoredResponses(ctx context.Context, requestJson []byte, storedRespFetcher stored_requests.Fetcher, bidderMap map[string]openrtb_ext.BidderName) (map[string]json.RawMessage, map[string]map[string]json.RawMessage, []error) {
+	impInfo, errs := parseImpInfo(requestJson)
```

Comment:
Could you explain again why we don't just call `exchange.parseImpInfo` instead of this function call that forces a copy/paste of the function? I know there's a goal of having the stored response code to be isolated from exchange, but I am confused as to why having the stored response code not be isolated is an issue.

Reply:
We want to have stored responses package outside of exchange and don't have dependencies from exchange. Isolated code is always good thing to have. In this case, unfortunately we need to duplicate this functionality to cut the dependency. Please refer to this comment: https://github.com/prebid/prebid-server/pull/2188#discussion_r834526341 
We also discussed it with Brian a lot offline.